### PR TITLE
JSON for Modern C++ 3.11.3

### PR DESCRIPTION
**Pull request by nlohmann**
# Bugs

- Adjust CMake files to accept `NEW` **CMake policies** up to CMake 3.14. This fixes a nasty deprecation warning that "Compatibility with CMake < 3.5 will be removed from a future version of CMake". #4076 #4112
- Fix CMake header path in install with custom `CMAKE_INSTALL_INCLUDEDIR`. #4194
- Add missing `<numeric>` header include. #3717 #3718 #3719
- Replace uses of `INT_MIN`/`INT_MAX`, etc. with `std::numeric_limits` and consistently use `std`-namespaced integer types to make library work with never GCC versions. #3722 #3723
- Add missing files (`json_fwd.hpp` and Bazel build files) to release artifact `include.zip`. #3727 #3728
- Fix 'declaration hides global declaration' warning. #3751
- Fix natvis XML. #3858 #3863
- Fix warning about moved from object. #3805 #3889
- Remove a magic number to fix a warning. #3837 #3888
- Fix debug pretty-printer by checking if match is valid before accessing group. #3919 #3920
- Fix custom allocators by defining missing `rebind` type. #3895 #3927
- Prevent memory leak when exception is thrown in `adl_serializer::to_json` #3881 #3901
- Fix Clang-Tidy warnings. #4047
- Fix init-list construction when `size_type` is not `int`. #4140
- Fix deprecation warning "identifier `_json` preceded by whitespace in a literal operator declaration". #4129 #4161
- Fix compile error with `_HAS_STATIC_RTTI=0`. #4046
- Fix char_traits deprecation warning "`char_traits<unsigned char>` is deprecated: `char_traits<T>` for `T` not equal to `char`, `wchar_t`, `char8_t`, `char16_t` or `char32_t` is non-standard". #4163 #4179

# Feature

- Allow [**custom base class**](https://json.nlohmann.me/api/basic_json/json_base_class_t/) as node customization point. This adds an additional template parameter which allows to set a custom base class for `nlohmann::json`. This class serves as an extension point and allows to add functionality to json node. Examples for such functionality might be metadata or additional member functions (e.g., visitors) or other application specific code. By default the parameter is set to `void` and an empty base class is used. In this case the library behaves as it already did. #3110
- Add more specific parse error message when attempting to **parse empty input**. #4037 #4180
- Add serialization-only user defined type [**macros**](https://json.nlohmann.me/features/macros/) (`NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_ONLY_SERIALIZE` and `NLOHMANN_DEFINE_TYPE_INTRUSIVE_ONLY_SERIALIZE`). #3816
- Add **Bazel** build support. If you are using [Bazel](https://bazel.build/) you can simply reference this repository using `http_archive` or `git_repository` and depend on `@nlohmann_json//:json`. #3709
- Support Apple's Swift Package Manager. #4010

# CI

- Use official [Clang](https://github.com/silkeh/docker-clang)/[GCC](https://hub.docker.com/_/gcc) Docker containers in the CI. #3703
- Add workflow to check if the source files area amalgamated. #3693
- Fix CI build. #3724 #3862 #3978 #3985 #4025 #4083 #4160 #4196 #4215
- Use Clang 15. #3822 #3876
- Add [CIFuzz](https://google.github.io/oss-fuzz/getting-started/continuous-integration/) CI GitHub action. #3845
- Fix MinGW script and CI. #3892 #4175
- Removed lgtm and DroneCI and added [Cirrus CI](https://cirrus-ci.org). #3890 #3906 #3935 #3937
- Set minimal permissions to Github Workflows. #3971 #3972
- Refactor amalgamation workflow to avoid dangerous use of pull_request_target. #3945 #3969

# Documentation

- Add [**migration guide**](https://json.nlohmann.me/integration/migration_guide/) to replace deprecated library functions. #3702 #3887
- Add dark mode toggle to documentation #3726
- Fix typos in documentation. #3748 #3767 #3902 #3932 #3951 #4109 #4126 #4143 #4149 #4159 #4173
- Update Codacy link. #3740
- Bump documentation tool versions. #3781 #3872 #3891 #3934
- Add vcpkg port version badge. #3988
- Add to `CONTRIBUTING.md` that `make pretty` is required for test updates. #4045
- Use template get instead of get in examples. #3827 #4038 #4039
- Capture exceptions by `const&` in docs. #4099
- Fix source highlighting in user defined type macros docs. #4169

# Tests

- Use `std::ranges::equals` for range comparisons in test case. #3927 #3950
- Add more algorithm tests to `unit-algorithm.cpp`. #4044

### CODE REVIEW

File: include/nlohmann/adl_serializer.hpp [hunk 1]
```diff
@@ -1,8 +1,9 @@
 //     __ _____ _____ _____
 //  __|  |   __|     |   | |  JSON for Modern C++
-// |  |  |__   |  |  | | | |  version 3.11.2
+// |  |  |__   |  |  | | | |  version 3.11.3
 // |_____|_____|_____|_|___|  https://github.com/nlohmann/json
 //
+// SPDX-FileCopyrightText: 2013-2022 Niels Lohmann <https://nlohmann.me>
```
Comment (gregmarr) [inline include/nlohmann/adl_serializer.hpp]: Did you intend to add this before the `-2023` version?

File: include/nlohmann/detail/json_custom_base_class.hpp [hunk 2]
```diff
@@ -1,3 +1,11 @@
+//     __ _____ _____ _____
+//  __|  |   __|     |   | |  JSON for Modern C++
+// |  |  |__   |  |  | | | |  version 3.11.3
+// |_____|_____|_____|_|___|  https://github.com/nlohmann/json
+//
+// SPDX-FileCopyrightText: 2013-2022 Niels Lohmann <https://nlohmann.me>
```
Comment (gregmarr) [inline include/nlohmann/detail/json_custom_base_class.hpp]: No 2023 here.

File: tests/src/unit-custom-base-class.cpp [hunk 3]
```diff
@@ -1,31 +1,11 @@
...
-OUT OF OR IN CONNECTION WITH THE SOFTWARE  OR THE USE OR OTHER DEALINGS IN THE
-SOFTWARE.
-*/
+//     __ _____ _____ _____
+//  __|  |   __|     |   | |  JSON for Modern C++ (supporting code)
+// |  |  |__   |  |  | | | |  version 3.11.3
+// |_____|_____|_____|_|___|  https://github.com/nlohmann/json
+//
+// Copyright (c) 2013-2019 Niels Lohmann <http://nlohmann.me>.
+// SPDX-FileCopyrightText: 2013-2022 Niels Lohmann <https://nlohmann.me>
```
Comment (gregmarr) [inline tests/src/unit-custom-base-class.cpp]: No 2023 here

File: include/nlohmann/adl_serializer.hpp (hunk 1 above)
Comment (nlohmann) [reply include/nlohmann/adl_serializer.hpp]: Good catch!

**Review by gregmarr — APPROVED** [review]

**Pull request merged**